hypertrace-trace-enricher/hypertrace-trace-enricher-api/src/main/java/org/hypertrace/traceenricher/trace/util/GraphBuilderUtil.java

```diff
@@ -7,34 +7,55 @@
 public class GraphBuilderUtil {
   private static final Logger LOG = LoggerFactory.getLogger(GraphBuilderUtil.class);
 
-  /**
-   * optimistic method of comparing two trace for considering rebuilding of entire graph structure.
-   */
-  static boolean isSameStructuredTrace(StructuredTrace cachedTrace, StructuredTrace trace) {
+  /** Check if the new trace is a different trace */
+  static boolean isDifferentTrace(StructuredTrace cachedTrace, StructuredTrace trace) {
     if (cachedTrace == null || trace == null) {
       LOG.debug("Cached and Input trace are not same. Reason: one of the input is null");
-      return false;
+      return true;
     }
-    // is processed and cached are same trace?
     if (!cachedTrace.getCustomerId().equals(trace.getCustomerId())
         || !cachedTrace.getTraceId().equals(trace.getTraceId())) {
       LOG.debug(
-          "Cached and Input trace are not same. Reason: doesn't not match either traceId or tenantId");
-      return false;
+          "Cached and Input trace are not same. Reason: doesn't match either traceId or tenantId");
+      return true;
     }
+    return false;
+  }
+
+  static boolean isStructuredTraceChanged(StructuredTrace cachedTrace, StructuredTrace trace) {
+    return isDifferentTrace(cachedTrace, trace)
+        || isTraceEventsChanged(cachedTrace, trace)
+        || isTraceEntitiesChanged(cachedTrace, trace);
+  }
 
-    // trace internally changed (full trace comparision is costly, so we are doing only with
+  /** Check if the events or theirs edges has changed */
+  static boolean isTraceEventsChanged(StructuredTrace cachedTrace, StructuredTrace trace) {
+    // trace events internally changed (full trace comparison is costly, so we are doing only with
     // required fields)
-    if (cachedTrace.getEntityList().size() != trace.getEntityList().size()
+    if (null == cachedTrace
```

Comment:
Shall we replace this with isDifferentTrace only?

Reply:
Done